lib/actions/preload.js

```diff
@@ -0,0 +1,426 @@
+				);
+			}
+			return progressBar.update({ percentage: event.percentage });
+		};
+
+		const spinners = {};
+
+		const spinnerHandler = function(event) {
+			let spinner = spinners[event.name];
+			if (!spinner) {
+				spinner = spinners[event.name] = new visuals.Spinner(event.name);
+			}
+			if (event.action === 'start') {
+				return spinner.start();
+			} else {
+				console.log();
+				return spinner.stop();
+			}
+		};
+
+		options.commit = isCurrent(options.commit) ? 'latest' : options.commit;
+		options.image = params.image;
+		options.appId = options.app;
+		delete options.app;
+
+		options.splashImage = options['splash-image'];
+		delete options['splash-image'];
+
+		options.dontCheckArch = options['dont-check-arch'] || false;
+		delete options['dont-check-arch'];
+		if (options.dontCheckArch && !options.appId) {
+			exitWithExpectedError(
+				'You need to specify an app id if you disable the architecture check.',
+			);
+		}
+
+		options.pinDevice = options['pin-device-to-release'] || false;
+		delete options['pin-device-to-release'];
+
+		if (Array.isArray(options['add-certificate'])) {
+			certificates = options['add-certificate'];
+		} else if (options['add-certificate'] === undefined) {
+			certificates = [];
+		} else {
+			certificates = [options['add-certificate']];
+		}
+		for (let certificate of certificates) {
+			if (!certificate.endsWith('.crt')) {
+				exitWithExpectedError('Certificate file name must end with ".crt"');
+			}
+		}
+
+		// Get a configured dockerode instance
+		return dockerUtils.getDocker(options).then(function(docker) {
+			const preloader = new balenaPreload.Preloader(
+				balena,
+				docker,
+				options.appId,
+				options.commit,
+				options.image,
+				options.splashImage,
+				options.proxy,
+				options.dontCheckArch,
+				options.pinDevice,
+				certificates,
+			);
+
+			let gotSignal = false;
+
+			nodeCleanup(function(_exitCode, signal) {
+				if (signal) {
+					gotSignal = true;
+					nodeCleanup.uninstall(); // don't call cleanup handler again
+					preloader.cleanup().then(() => {
+						// calling process.exit() won't inform parent process of signal
+						process.kill(process.pid, signal);
+					});
+					return false;
+				}
+			});
+
+			if (process.env.DEBUG) {
+				preloader.stderr.pipe(process.stderr);
+			}
+
+			preloader.on('progress', progressHandler);
+			preloader.on('spinner', spinnerHandler);
+
+			return new Promise(function(resolve, reject) {
+				preloader.on('error', reject);
+
+				return preloader
+					.prepare()
+					.then(() =>
+						// If no appId was provided, show a list of matching apps
+						Promise.try(function() {
+							if (!preloader.appId) {
+								return selectApplication(
+									preloader.config.deviceType,
+								).then(application => preloader.setApplication(application));
+							}
+						}),
+					)
+					.then(() => {
+						// Use the commit given as --commit or show an interactive commit selection menu
+						if (options.commit) {
```

Comment:
This used to be in a `Promise.try` - I removed it as it was unnecessary and doing so plays better with types